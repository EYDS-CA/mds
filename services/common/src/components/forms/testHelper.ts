export const inputMeta = {
    touched: false,
    error: false,
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: false,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: false,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: false,
};

export const errorInputMeta = {
    touched: true,
    error: "Error message",
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: true,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: true,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: true,
}

// needs a value and a name
export const inputProps = {
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onDragStart: jest.fn(),
    onDrop: jest.fn(),
    onFocus: jest.fn()
}