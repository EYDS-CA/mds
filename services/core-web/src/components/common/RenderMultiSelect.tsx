import React, { FC, useRef } from "react";
import { Form } from "@ant-design/compatible";
import "@ant-design/compatible/assets/index.css";
import { Select } from "antd";
import { caseInsensitiveLabelFilter } from "@common/utils/helpers";
import { debounce} from "lodash";
import { WrappedFieldMetaProps } from "redux-form";
import { IOption } from "@mds/common/interfaces";

/**
 * @constant RenderSelect - Ant Design `Select` component for redux-form - used for small data sets that (< 100);
 */
interface RenderMultiSelectProps {
  id: string,
  input: any,
  placeholder?: string,
  label?: string,
  meta: WrappedFieldMetaProps,
  data: IOption[],
  filterOption: (input: any, option: any) => void | boolean,
  disabled: boolean,
  onSearch: (value: any) => void,
  isModal: boolean,
};

export const RenderMultiSelect: FC<RenderMultiSelectProps> = ({
  placeholder = "",
  label = "",
  data = [],
  disabled = false,
  meta = { touched: false},
  onSearch = () => { },
  filterOption = false,
  isModal = false,
  ...rest
}) => {
  const debouncedSearch = useRef(debounce(onSearch, 500)).current;

  const extraProps = isModal ? null : { getPopupContainer: (trigger) => trigger.parentNode };
  return (
    <div>
      <Form.Item
        label={label}
        validateStatus={
          meta.touched
            ? (meta.error && "error") || (meta.warning && "warning")
            : ""
        }
        help={
          meta.touched &&
          ((meta.error && <span>{meta.error}</span>) ||
            (meta.warning && <span>{meta.warning}</span>))
        }
      >
        <Select
          virtual={false}
          disabled={!data || disabled}
          mode="multiple"
          size="small"
          placeholder={placeholder}
          id={rest.id}
          onSearch={debouncedSearch}
          value={rest.input.value ? rest.input.value : undefined}
          onChange={rest.input.onChange}
          filterOption={filterOption || caseInsensitiveLabelFilter}
          showArrow
          {...extraProps}
        >
          {data &&
            data.map(({ value, label, tooltip }) => (
              <Select.Option key={value} value={value} title={tooltip}>
                {label}
              </Select.Option>
            ))}
        </Select>
      </Form.Item>
    </div>
  );
};

export default RenderMultiSelect;
