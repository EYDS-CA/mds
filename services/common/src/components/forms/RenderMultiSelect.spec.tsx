import React from "react";
import { render } from "@testing-library/react";
import RenderMultiSelect from "./RenderMultiSelect";
import { inputMeta, inputProps } from "./testHelper";

describe("RenderMultiSelect", () => {
    it("renders properly", () => {
        const { container } = render(<RenderMultiSelect
            id="multiInput"
            data={[
                { label: "Opt 1", value: "opt1" },
                { label: "Opt 2", value: "opt2" },
                { label: "Opt 3", value: "opt3" }
            ]}
            input={{ value: ["opt1", "opt2"], name: "multiInput", ...inputProps }}
            meta={inputMeta}
        />);
        expect(container).toMatchSnapshot();
    })
})