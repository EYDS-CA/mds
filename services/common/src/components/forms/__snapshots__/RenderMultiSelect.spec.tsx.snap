// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderMultiSelect renders properly 1`] = `
<div>
  <div
    class="form-multiselect"
  >
    <div
      class="ant-form-item ant-form-item-with-help"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-sm ant-select-in-form-item ant-select-multiple ant-select-show-arrow ant-select-show-search"
                style="width: 100%;"
              >
                <div
                  class="ant-select-selector"
                >
                  <div
                    class="ant-select-selection-overflow"
                  >
                    <div
                      class="ant-select-selection-overflow-item"
                      style="opacity: 1;"
                    >
                      <span
                        class="ant-select-selection-item"
                        title="Opt 1"
                      >
                        <span
                          class="ant-select-selection-item-content"
                        >
                          Opt 1
                        </span>
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-item-remove"
                          style="user-select: none;"
                          unselectable="on"
                        >
                          <span
                            aria-label="close"
                            class="anticon anticon-close"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close"
                              fill="currentColor"
                              fill-rule="evenodd"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                    <div
                      class="ant-select-selection-overflow-item"
                      style="opacity: 1;"
                    >
                      <span
                        class="ant-select-selection-item"
                        title="Opt 2"
                      >
                        <span
                          class="ant-select-selection-item-content"
                        >
                          Opt 2
                        </span>
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-item-remove"
                          style="user-select: none;"
                          unselectable="on"
                        >
                          <span
                            aria-label="close"
                            class="anticon anticon-close"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close"
                              fill="currentColor"
                              fill-rule="evenodd"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                    <div
                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                      style="opacity: 1;"
                    >
                      <div
                        class="ant-select-selection-search"
                        style="width: 0px;"
                      >
                        <input
                          aria-activedescendant="multiInput_list_0"
                          aria-autocomplete="list"
                          aria-controls="multiInput_list"
                          aria-haspopup="listbox"
                          aria-owns="multiInput_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="multiInput"
                          readonly=""
                          role="combobox"
                          style="opacity: 0;"
                          type="search"
                          unselectable="on"
                          value=""
                        />
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-search-mirror"
                        >
                          
                          Â 
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            style="display: flex; flex-wrap: nowrap;"
          >
            <div
              class="ant-form-item-explain ant-form-item-explain-connected"
              id="multiInput_help"
              role="alert"
            >
              <div
                class=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
